File: scripts/test_history.mjs
/**
 * 已完成历史功能 —— 局部逻辑回归测试（Node，标记"未写生产"）
 *
 * 遵循 AGENTS.md 铁律一/二：
 *   - 不触碰任何生产数据 / 不连数据库，纯逻辑断言。
 *   - 复用真实 state.js 的 sortTodos（已完成沉底），并复刻 app.js 里
 *     主列表过滤 / 历史列表过滤+排序 / 顶栏徽标计数的断言逻辑，
 *     验证"主页只渲染未完成、已完成进历史页"的拆分行为正确。
 *
 * 注意：本测试验证的是「数据变换逻辑」，不是真机 UI 交互（弹层开合、动画、
 * 长按手势、对方实时同步体验等需真机/Playwright，见交付报告硬限制）。
 *
 * 用法：node scripts/test_history.mjs
 */
import { sortTodos } from '../public/js/state.js';

let pass = 0;
let fail = 0;
function check(name, cond, detail = '') {
  if (cond) {
    pass++;
    console.log(`  OK  ${name}`);
  } else {
    fail++;
    console.log(`  FAIL ${name} ${detail}`);
  }
}

// ===== 复刻 app.js 的真实断言逻辑（与源码保持语义一致）=====
// 主页：只渲染未完成
const filterActive = (list) => list.filter((t) => !t.completed);
// 历史：只渲染已完成且未软删除，按 completedAt 倒序
const filterHistory = (list) =>
  list
    .filter((t) => t.completed && !t.deleted_at)
    .sort((a, b) => {
      const ta = a.completedAt || a.createdAt || '';
      const tb = b.completedAt || b.createdAt || '';
      return tb.localeCompare(ta);
    });
// 顶栏徽标：已完成数量
const badgeCount = (list) => list.filter((t) => t.completed && !t.deleted_at).length;

const now = Date.now();
const iso = (offsetMs) => new Date(now + offsetMs).toISOString();

console.log('== 场景1：添加 1 条未完成 → 主列表显示，历史列表没有 ==');
{
  const todos = [{ id: 'a', text: '买菜', completed: false, createdAt: iso(0) }];
  const main = filterActive(todos);
  const history = filterHistory(todos);
  check('主列表含该待办', main.length === 1 && main[0].id === 'a');
  check('历史列表为空', history.length === 0);
  check('徽标为 0（隐藏态逻辑）', badgeCount(todos) === 0);
}

console.log('== 场景2：标记完成 → 主列表消失，历史列表出现，badge+1 ==');
{
  const todos = [{ id: 'a', text: '买菜', completed: true, createdAt: iso(0), completedAt: iso(100) }];
  const main = filterActive(todos);
  const history = filterHistory(todos);
  check('主列表为空', main.length === 0);
  check('历史列表出现该条', history.length === 1 && history[0].id === 'a');
  check('徽标 = 1', badgeCount(todos) === 1);
}

console.log('== 场景3：取消完成（历史页长按）→ 历史消失，主列表出现 ==');
{
  const todos = [{ id: 'a', text: '买菜', completed: false, createdAt: iso(0) }];
  const main = filterActive(todos);
  const history = filterHistory(todos);
  check('主列表重新出现该条', main.length === 1 && main[0].id === 'a');
  check('历史列表再次为空', history.length === 0);
  check('徽标回到 0', badgeCount(todos) === 0);
}

console.log('== 场景4：badge 数量实时同步（多已完成）==');
{
  const todos = [
    { id: 'a', completed: true, completedAt: iso(100) },
    { id: 'b', completed: false },
    { id: 'c', completed: true, completedAt: iso(300) },
    { id: 'd', completed: true, completedAt: iso(200) },
    { id: 'e', completed: true, completedAt: iso(50), deleted_at: iso(400) }, // 软删除不计数
  ];
  check('徽标 = 3（排除未完成与软删除）', badgeCount(todos) === 3);
  const history = filterHistory(todos);
  check('历史列表 = 3 条', history.length === 3);
  // 倒序校验：completedAt 300 > 200 > 100
  check('排序：最近完成在前 (c,d,a)', history.map((t) => t.id).join(',') === 'c,d,a');
  // 软删除的 e 不出现在历史
  check('软删除项不进历史', !history.some((t) => t.id === 'e'));
  // 主列表只含未完成 b
  check('主列表只含未完成 b', filterActive(todos).map((t) => t.id).join(',') === 'b');
}

console.log('== 场景4b：badge 上限 99+ ==');
{
  const todos = Array.from({ length: 150 }, (_, i) => ({ id: 'x' + i, completed: true, completedAt: iso(i) }));
  // badgeCount 真实值=150，UI 层会把 >99 显示成 "99+"，这里验证计数逻辑正确
  check('计数逻辑 = 150（UI 负责截断 99+）', badgeCount(todos) === 150);
}

console.log('== 场景5：sortTodos 已完成沉底（主列表过滤前的底层保证）==');
{
  const list = [
    { id: 'done1', completed: true, createdAt: iso(300) },
    { id: 'todo1', completed: false, createdAt: iso(100) },
    { id: 'todo2', completed: false, createdAt: iso(200) },
  ];
  const sorted = sortTodos(list);
  check('已完成沉到最后', sorted[sorted.length - 1].id === 'done1');
  check('未完成在前且按创建时间倒序', sorted[0].id === 'todo2' && sorted[1].id === 'todo1');
}

console.log('== 场景5b：主列表过滤后“完成即消失”的视觉前提 ==');
{
  // 模拟实时回推：某条从未完成→已完成，setTodos 触发 render→filterActive
  const before = [{ id: 'a', completed: false }];
  const after = before.map((t) => (t.id === 'a' ? { ...t, completed: true, completedAt: iso(0) } : t));
  check('完成前在主列表', filterActive(before).length === 1);
  check('完成后移出主列表', filterActive(after).length === 0);
  check('完成后进入历史列表', filterHistory(after).length === 1);
}

console.log('========================================');
console.log(`通过: ${pass}  失败: ${fail}`);
process.exit(fail === 0 ? 0 : 1);

File: scripts/test_history.py
"""
已完成历史功能 E2E 测试（真实 UI / 双账号之一：小宝宝）

覆盖：主页只渲染未完成、完成即移出主页、历史弹层出现、顶栏徽标计数、
历史页取消完成回流到主页、弹层开合。

遵循 AGENTS.md：测试待办用 "E2E-测试-" 前缀，测完彻底清理（回收站+彻底删除）。
注意：用精确 data-id 匹配，避免 has_text 对长中文/特殊字符的误匹配。
用法：先 node scripts/serve.mjs，再 python3 scripts/test_history.py
"""
import sys
from playwright.sync_api import sync_playwright

BASE = "http://localhost:3000"
TEST_TEXT = "E2E-测试-历史页"
errors = []
results = {"pass": 0, "fail": 0}


def check(name, cond, detail=""):
    if cond:
        results["pass"] += 1
        print(f"  OK {name}", flush=True)
    else:
        results["fail"] += 1
        print(f"  FAIL {name} {detail}", flush=True)


def add_todo(page, text):
    page.locator('#fabBtn').click()
    page.wait_for_selector('#addPanel.add-panel--show', timeout=5000)
    page.fill('#todoInput', text)
    page.locator('#addBtn').click()
    page.wait_for_selector(f'#todoList .todo:has-text("{text}")', timeout=8000)


def cleanup(page, data_id):
    """按 data-id 定位，右键→删除→等待撤销 toast 消失→回收站→彻底删除"""
    sel = f'.todo[data-id="{data_id}"]'
    try:
        page.locator(sel).first.click(button='right')
        page.wait_for_selector('.action-sheet__icon-btn[aria-label="删除"]', timeout=5000)
        page.locator('.action-sheet__icon-btn[aria-label="删除"]').click()
        page.wait_for_timeout(5500)
        page.locator(sel).first.wait_for(state='detached', timeout=5000)

        page.locator('.topbar__avatar').click(button='right')
        page.wait_for_selector('.account-menu__item', timeout=5000)
        page.locator('.account-menu__item', has_text='回收站').first.click()
        page.wait_for_selector('.trash-item', timeout=8000)
        purge = page.locator(f'.trash-item[data-id="{data_id}"] .trash-item__btn--purge')
        if purge.count() == 0:
            purge = page.locator('.trash-item', has_text=TEST_TEXT).first.locator('.trash-item__btn--purge')
        purge.click()
        page.wait_for_timeout(500)
        armed = page.locator('.trash-item__btn--armed')
        if armed.count() >= 1:
            armed.first.click()
            page.wait_for_timeout(3000)
    except Exception as e:
        print(f"  [warn] 清理异常: {e}", flush=True)


with sync_playwright() as p:
    browser = p.chromium.launch(headless=True)
    page = browser.new_page()
    page.set_default_timeout(15000)
    page.on("console", lambda m: errors.append(f"[{m.type}] {m.text}") if m.type == "error" else None)
    page.on("pageerror", lambda e: errors.append(f"[pageerror] {e}"))

    print("== 1. 登录 ==", flush=True)
    page.goto(f"{BASE}/login.html", wait_until="domcontentloaded")
    page.wait_for_selector('#username', timeout=15000)
    page.fill('#username', '小宝宝')
    page.fill('#password', '5201314')
    page.click('#submitBtn')
    page.wait_for_selector('.topbar__avatar', timeout=15000)
    page.wait_for_timeout(1500)

    print("== 2. 添加测试待办 ==", flush=True)
    add_todo(page, TEST_TEXT)
    item = page.locator('#todoList .todo', has_text=TEST_TEXT).first
    data_id = item.get_attribute('data-id')
    check("主列表含新待办", page.locator(f'#todoList .todo[data-id="{data_id}"]').count() == 1)
    check("历史入口存在", page.locator('#historyEntry').count() == 1)
    check("历史徽标元素存在", page.locator('#historyBadge').count() == 1)

    print("== 3. 标记完成 → 移出主页 + 进入历史 ==", flush=True)
    page.locator(f'#todoList .todo[data-id="{data_id}"] .todo__check').click()
    page.wait_for_timeout(1000)
    check("主页已移除该待办", page.locator(f'#todoList .todo[data-id="{data_id}"]').count() == 0)
    page.locator('#historyEntry').click()
    page.wait_for_selector('#historyModal:not(.hidden)', timeout=5000)
    check("历史弹层已打开", True)
    check("历史列表出现该待办", page.locator(f'#historyList .todo[data-id="{data_id}"]').count() == 1)

    print("== 4. 关闭弹层 ==", flush=True)
    page.locator('#historyModalClose').click()
    page.wait_for_function("document.getElementById('historyModal').classList.contains('hidden')", timeout=5000)
    check("历史弹层已关闭", True)

    print("== 5. 历史页取消完成 → 回流到主页 ==", flush=True)
    page.locator('#historyEntry').click()
    page.locator(f'#historyList .todo[data-id="{data_id}"]').first.wait_for(state='attached', timeout=5000)
    page.locator(f'#historyList .todo[data-id="{data_id}"]').first.click(button='right')
    page.wait_for_selector('.action-sheet__icon-btn[aria-label="取消完成"]', timeout=5000)
    page.locator('.action-sheet__icon-btn[aria-label="取消完成"]').first.click()
    # 等待历史列表移除该精确 data-id 的节点
    try:
        page.locator(f'#historyList .todo[data-id="{data_id}"]').first.wait_for(state='detached', timeout=6000)
        history_gone = True
    except Exception:
        history_gone = page.locator(f'#historyList .todo[data-id="{data_id}"]').count() == 0
    check("历史列表已移除该待办", history_gone)
    check("主页重新出现该待办", page.locator(f'#todoList .todo[data-id="{data_id}"]').count() == 1)
    # 关闭弹层（若仍开）
    if not page.evaluate("document.getElementById('historyModal').classList.contains('hidden')"):
        page.locator('#historyModalClose').click()
        page.wait_for_timeout(400)

    print("== 6. 清理测试数据 ==", flush=True)
    cleanup(page, data_id)
    page.wait_for_timeout(1500)
    # 主列表与历史列表都不应再出现该 data-id
    left_main = page.locator(f'#todoList .todo[data-id="{data_id}"]').count()
    left_hist = page.locator(f'#historyList .todo[data-id="{data_id}"]').count()
    check("测试待办已彻底清除(主列表)", left_main == 0, f"main={left_main}")
    check("测试待办已彻底清除(历史列表)", left_hist == 0, f"hist={left_hist}")

    print("== 7. 页面报错检查 ==", flush=True)
    real = [e for e in errors if "Failed to fetch" not in e and "net::" not in e and "favicon" not in e]
    check("无模块级报错", len(real) == 0, f"错误数 {len(real)}")
    for e in real[:5]:
        print(f"    {e}", flush=True)

    browser.close()

print("=" * 40, flush=True)
print(f"通过: {results['pass']}  失败: {results['fail']}", flush=True)
sys.exit(0 if results["fail"] == 0 else 1)
